src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -1523,6 +1523,137 @@ void C2_MacroAssembler::vinsert(BasicType typ, XMMRegister dst, XMMRegister src,
+                                          Register mask, Register midx, Register rtmp, int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
+  if (elem_bt == T_SHORT) {
+    Label case0, case1, case2, case3;
+    Label* larr[] = { &case0, &case1, &case2, &case3 };
+    for (int i = 0; i < 4; i++) {
+      bt(mask, midx);
+      jccb(Assembler::carryClear, *larr[i]);
+      movl(rtmp, Address(idx_base, i*4));
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
+      bind(*larr[i]);
+      incq(midx);
+    }
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    Label case0, case1, case2, case3, case4, case5, case6, case7;
+    Label* larr[] = { &case0, &case1, &case2, &case3, &case4, &case5, &case6, &case7 };
+    for (int i = 0; i < 8; i++) {
+      bt(mask, midx);
+      jccb(Assembler::carryClear, *larr[i]);
+      movl(rtmp, Address(idx_base, i*4));
+      pinsrb(dst, Address(base, rtmp), i);
+      bind(*larr[i]);
+      incq(midx);
+    }
+  }
+}
+
+void C2_MacroAssembler::vpgather8b_masked_offset(BasicType elem_bt, XMMRegister dst, Register base, Register idx_base,
+                                                 Register offset, Register mask, Register midx, Register rtmp, int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
+  if (elem_bt == T_SHORT) {
+    Label case0, case1, case2, case3;
+    Label* larr[] = { &case0, &case1, &case2, &case3 };
+    for (int i = 0; i < 4; i++) {
+      bt(mask, midx);
+      jccb(Assembler::carryClear, *larr[i]);
+      movl(rtmp, Address(idx_base, i*4));
+      addl(rtmp, offset);
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
+      bind(*larr[i]);
+      incq(midx);
+    }
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    Label case0, case1, case2, case3, case4, case5, case6, case7;
+    Label* larr[] = { &case0, &case1, &case2, &case3, &case4, &case5, &case6, &case7 };
+    for (int i = 0; i < 8; i++) {
+      bt(mask, midx);
+      jccb(Assembler::carryClear, *larr[i]);
+      movl(rtmp, Address(idx_base, i*4));
+      addl(rtmp, offset);
+      pinsrb(dst, Address(base, rtmp), i);
+      bind(*larr[i]);
+      incq(midx);
+    }
+  }
+}
+#endif // _LP64
+
+void C2_MacroAssembler::vpgather8b(BasicType elem_bt, XMMRegister dst, Register base, Register idx_base, Register rtmp, int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
+  if (elem_bt == T_SHORT) {
+    for (int i = 0; i < 4; i++) {
+      movl(rtmp, Address(idx_base, i*4));
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
+    }
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    for (int i = 0; i < 8; i++) {
+      movl(rtmp, Address(idx_base, i*4));
+      pinsrb(dst, Address(base, rtmp), i);
+    }
+  }
+}
+
+void C2_MacroAssembler::vpgather8b_offset(BasicType elem_bt, XMMRegister dst, Register base, Register idx_base,
+                                          Register offset, Register rtmp, int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
```

Comment:
Why the vpxor here and in vpgather8b. These are not masked gathers.

Reply:
This is not a masked operation so every lane of dst will be written through pinsrw/pinsrb. An vpxor before is not required.

Reply:
> This is not a masked operation so every lane of dst will be written through pinsrw/pinsrb. An vpxor before is not required.

xor here clears the intermediate vector after each iteration,  this is eventually ORs with destination.

Checkout line https://github.com/openjdk/jdk/pull/16354/files/86783403c453d329e33d94f787a5709ec35f7099#diff-318d0e76b9a97e8cf8936be1de34e52735c4d947a77cac38babbbf9a081d16fcR1644

https://github.com/openjdk/jdk/pull/16354/files/86783403c453d329e33d94f787a5709ec35f7099#diff-318d0e76b9a97e8cf8936be1de34e52735c4d947a77cac38babbbf9a081d16fcR1651